src/core/catalog.js

```diff
@@ -486,17 +486,17 @@ class Catalog {
         return shadow(this, "optionalContentConfig", null);
       }
       const groups = [];
-      const groupRefs = new RefSet();
+      const groupRefCache = new RefSetCache();
       // Ensure all the optional content groups are valid.
       for (const groupRef of groupsData) {
-        if (!(groupRef instanceof Ref) || groupRefs.has(groupRef)) {
+        if (!(groupRef instanceof Ref) || groupRefCache.has(groupRef)) {
           continue;
         }
-        groupRefs.put(groupRef);
-
-        groups.push(this.#readOptionalContentGroup(groupRef));
+        const group = this.#readOptionalContentGroup(groupRef);
+        groups.push(group);
+        groupRefCache.put(groupRef, group);
       }
-      config = this.#readOptionalContentConfig(defaultConfig, groupRefs);
+      config = this.#readOptionalContentConfig(defaultConfig, groupRefCache);
       config.groups = groups;
```

Comment:
Do we actually need the separate `groups` variable now, or could we just do the following here?
```suggestion
      config.groups = [...groupRefCache];
```